File: src/main/java/cspom/dimacs/CNFParser.java
package cspom.dimacs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import cspom.CSPOM;
import cspom.CSPParseException;
import cspom.compiler.PredicateParseException;
import cspom.variable.CSPOMVariable;
import cspom.variable.UnknownBooleanDomain$;

public final class CNFParser {
    private final CSPOM problem;

    private static final Pattern PARAMETER = Pattern
            .compile("^p cnf (\\d+) (\\d+)$");
    private static final Pattern VAR = Pattern.compile("(-?\\d+)");

    public CNFParser(final CSPOM problem) {
        this.problem = problem;
    }

    public void parse(final InputStream is) throws IOException,
            CSPParseException {
        final BufferedReader reader = new BufferedReader(new InputStreamReader(
                is));

        List<Integer> currentClause = new ArrayList<Integer>();
        boolean parameter = false;
        for (String line = reader.readLine(); line != null; line = reader
                .readLine()) {
            if (line.startsWith("c")) {
                continue;
            }
            if (!parameter) {
                final Matcher matcher = PARAMETER.matcher(line);
                if (matcher.matches()) {
                    parameter = true;
                    final int nbVars = Integer.valueOf(matcher.group(1));
                    for (int i = 1; i <= nbVars; i++) {
                        problem.addVariable(new CSPOMVariable("V" + i,
                                UnknownBooleanDomain$.MODULE$));
                    }
                    continue;
                }
                throw new CSPParseException("Parameter line not found");
            }
            final Matcher varMatcher = VAR.matcher(line);
            while (varMatcher.find()) {
                final int var = Integer.valueOf(varMatcher.group(1));
                if (var == 0) {
                    try {
                        problem.ctr(clause(currentClause));
                    } catch (PredicateParseException e) {
                        throw new IllegalStateException(e);
                    }
                    currentClause.clear();
                } else {
                    currentClause.add(var);
                }
            }
        }
    }

    private String clause(final List<Integer> currentClause) {

        final StringBuilder clause = new StringBuilder();
        final StringBuilder parameters = new StringBuilder();
        parameters.append("or{");

        final Iterator<Integer> i = currentClause.iterator();

        for (;;) {
            int v = i.next();
            clause.append('V').append(Math.abs(v));
            if (v > 0) {
                parameters.append('0');
            } else {
                parameters.append('1');
            }

            if (!i.hasNext()) {
                return parameters.append("}(").append(clause.toString())
                        .append(')').toString();
            }
            clause.append(", ");
            parameters.append(", ");
        }

    }
}

File: src/main/scala/cspom/dimacs/CNFParser.scala
package cspom.dimacs;

import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.Iterator
import scala.io.Source
import scala.util.matching.Regex
import cspom.CSPOM
import cspom.CSPParseException
import cspom.CSPOMConstraint
import cspom.variable.CSPOMVariable
import CSPOM._
import cspom.variable.BoolVariable
import cspom.variable.CSPOMExpression
import scala.util.Try
import cspom.CSPOMGoal

final object CNFParser extends CSPOM.Parser {

  private val PARAMETER = new Regex("""^p cnf (\d+) (\d+)$""");
  private val VAR = new Regex("""(-?\d+)""");

  def apply(is: InputStream): Try[CSPOM] = Try {
    val reader = new BufferedReader(new InputStreamReader(is));

    val lines = Source.fromInputStream(is).getLines.filter(
      s => !(s startsWith "c") && !(s.trim.isEmpty))

    val PARAMETER(nbVars, nbClauses) = try lines.next
    catch {
      case e: Exception => throw new CSPParseException("Parameter line not found", e, -1)
    }

    var countClauses = 0
    val problem = CSPOM { implicit problem =>

      val (variables, ns) = (for (i <- 1 to nbVars.toInt) yield {
        new BoolVariable() withName ("V" + i)
      }).unzip

      var currentClause: List[Int] = Nil;

      for (line <- lines; value <- VAR.findAllIn(line).map(s => s.toInt)) {
        if (value == 0) {
          ctr(clause(currentClause, variables));
          countClauses += 1;
          currentClause = Nil;
        } else {
          currentClause ::= value;
        }
      }

      CSPOM.goal(CSPOMGoal.Satisfy)
    }

    require(countClauses == nbClauses.toInt)

    problem
  }

  private def clause(currentClause: List[Int], variables: IndexedSeq[CSPOMExpression[Boolean]]) = {
    require(!currentClause.contains(0))

    val (positive, negative) = currentClause.partition(_ > 0)

    /*
     * Variable indices starts at 1 in Dimacs format, 0 in IndexedSeq
     */

    CSPOMConstraint('clause)(positive.map(i => variables(i - 1)), negative.map(i => variables(-i - 1)))

  }
}
